crates/red_knot_python_semantic/resources/mdtest/exception/basic.md

```diff
@@ -41,11 +41,10 @@ except EXCEPTIONS as f:
 ## Dynamic exception types
 
 ```py
-# TODO: we should not emit these `call-possibly-unbound-method` errors for `tuple.__class_getitem__`
 def foo(
     x: type[AttributeError],
-    y: tuple[type[OSError], type[RuntimeError]],  # error: [call-possibly-unbound-method]
-    z: tuple[type[BaseException], ...],  # error: [call-possibly-unbound-method]
+    y: tuple[type[OSError], type[RuntimeError]],
+    z: tuple[type[BaseException], ...],
```

Comment:
I'm actually not sure the TODO comment is correct here. Our default `--target-version` is currently Python 3.8. This function will indeed fail on Python 3.8, since that's before PEP-585, and there's no `from __future__ import annotations` in this snippet.

Can you explain why these errors are going away here? Is it because we now only infer subscripts in annotations according to their type-expression semantics, whereas previously we inferred them according to their type-expression semantics _and_ their runtime semantics?

Reply:
Yes, it's because after this PR we only infer annotations using `infer_type_expression`, which currently only handles subscripting for a limited set of forms.

It's correct to only use `infer_type_expression` for annotations, but we need to improve and expand `infer_type_expression`. But I don't think a TODO here for these errors is really needed, since we already have todos to actually understand these annotations, and dealing with the possibly-unboundness will be necessary when we do that.

Reply:
> But I don't think a TODO here for these errors is really needed, since we already have todos to actually understand these annotations, and dealing with the possibly-unboundness will be necessary when we do that.

By "these annotations", you mean "annotations for function parameters"? We already understand `tuple[type[OSError], type[RuntimeError]]` in a type expression, if you run red-knot on this snippet:

```py
from typing_extensions import reveal_type

def foo() -> tuple[type[OSError], type[RuntimeError]]: ...
reveal_type(foo())
```

Reply:
Ah ok -- then I think the problem is just that the way we interpret this currently in a type expression doesn't actually go through a call to `tuple.__class_getitem__`, it's handled as a special form.

So maybe we do want a TODO for erroring on this, if in a non-deferred annotation, on an old Python version. I think this would kind of have to be implemented as a special-case diagnostic here. And I'm not sure how high a priority it is, since it's a disappearing legacy corner case: annotations are going to all-deferred, and only the oldest Python versions don't have `tuple.__getitem__`.

Reply:
Hmm, maybe we don't actually need a TODO, since once we understand generic types, we'll think everything that inherits from `Generic` has a `__class_getitem__` method, and `tuple` inherits from `Generic`[^1].

But on the other hand, we may want to implement some special casing for stdlib types that typeshed has as inheriting from `Generic` but are not _explicitly_ annotated as having `__class_getitem__` methods. For example: `memoryview` [is a generic type in typeshed](https://github.com/python/typeshed/blob/af4df4eb24f1cda42033f74440ba1117ce4c8f7e/stdlib/builtins.pyi#L840-L907) in all `sys.version_info` branches, but the `__class_getitem__` method has only been added in 3.14 (the new `__class_getitem__` method hasn't even been reflected in typeshed yet!), so attempting to subscript it will fail on Python <=3.13. Ideally we'd catch that error, and similar errors for other stdlib classes that have only had `__class_getitem__` methods added in newer Python versions, but are generic on all versions according to typeshed.

[^1]: Not in real life but at least according to typeshed

Reply:
> we may want to implement some special casing for stdlib types

Maybe, but IMO it is pretty far down on the priority list. Mypy does it, but pyright does not, and the direction of the future is that all annotations are deferred by default, which I think means not erroring on such cases.